### PR TITLE
gh-106690: Add a .coveragerc file to the CPython repository

### PR DESCRIPTION
**Pull request by ammaraskar**
This was requested by Terry Reedy, the coveragerc file is theirs. It ignores code blocks that don't really matter and raise false alarms in terms of missed coverage.

(Can be marked with skip issue and skip news as it pertains to the build process)

* Issue: gh-106690

### CODE REVIEW

**Comment by ammaraskar** [comment]
@terryjreedy Coverage went up +1.30% with this: https://codecov.io/gh/python/cpython/pull/8150

This is probably a good place to start a discussion on if there's any other common patterns in the python code that should also be excluded. 

Getting a review from someone knowledgeable with the build process like Brett or Zach would be good for this.

**Comment by ammaraskar** [comment]
Also, `branch = True` caused some reliably reproducible test errors, will make bpo bugs for them later after investigating.

**Comment by zware** [comment]
@ammaraskar Can you rebase this please?  I think this is good to merge.

**Comment by ammaraskar** [comment]
@zware Sorry for the late response. I needed some time to context-switch back to this, it didn't work last time I tried and I believe it's because codecov itself needs to be aware of the `.coveragerc` file and can't use it if it's not present in the root of the project. It seems to work at a quick glance:

* https://codecov.io/gh/python/cpython/src/60303afbe5c540190a9b9dc9935bdb1272e13b09/Lib/dis.py#L552

* https://codecov.io/gh/python/cpython/src/60303afbe5c540190a9b9dc9935bdb1272e13b09/Lib/tokenize.py#L681

* https://codecov.io/gh/python/cpython/src/60303afbe5c540190a9b9dc9935bdb1272e13b09/Lib/asyncio/__init__.py#L41

**Comment by csabella** [comment]
@zware, do you want to take another look before merging?  Thanks!

**Comment by zware** [comment]
I've not had a chance to swap this back into my brain lately and don't foresee such an opportunity in the near future.  I'm in favor of adding a `.coveragerc` even if it's not complete, to be iterated on as we go; if @terryjreedy is good with this one, so am I :)

**Comment by arhadthedev** [comment]
> I'm in favor of adding a .coveragerc even if it's not complete, to be iterated on as we go; if @terryjreedy is good with this one, so am I :)

Pinging @terryjreedy to resolve this PR stuck for five years total.

File: .coveragerc
Comment (terryjreedy) [inline .coveragerc:15]: Except for removal of branch coverage, which is maybe a separate issue, this new file matches what I use for IDLE currently.  But I think it needs an explanatory comment added.
```suggestion
    # Additions for IDLE:
    .*# htest #
```

**Comment by terryjreedy** [comment]
I requested this back when coverage was run with CI and, as I remember, it was proposed that decreased coverage should block merging.  Without exclusion of code that cannot be covered by unittests, improvements can decrease 'coverage'.

While removal of CI coverage and withdrawal of the gateway proposal removed he urgency of this patch, the accuracy point remains true.  My (biased) feeling is that this should still be added (with the comment I added).

It is unclear to me whether or when the addition would affect a particular coverage run.  The way I run IDLE coverage, it needs to be in the directory that contains the repository.  But others cannot, if needed, copy to another place without something in the repository to copy.

**Comment by erlend-aasland** [comment]
@arhadthedev: why was the `pending` label added? I see no reason why, and no accompanying comment was added with the label (as should have been done).

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Encouraged but not "GitHub approved" by Zach here: https://github.com/python/cpython/pull/8150#issuecomment-581497730

**Comment by arhadthedev** [comment]
> no accompanying comment was added with the label (as should have been done).

I agree; back then the reasoning seemed to be extremely obvious for me but now I see it held on incorrect assumptions.

> @arhadthedev: why was the pending label added?

There was no feedback from an author for 1.5 month after the addition of https://github.com/python/cpython/labels/DO-NOT-MERGE so I deemed the PR totally abandoned.

Also, I didn't know back then that stuff can totally fall through the cracks thus thinking that if a PR is not merged after many years, its approach is incorrect and needs to be reworked. As a result, I concluded that the PR is likely unsalvageable but deserves the last chance through addition of https://github.com/python/cpython/labels/pending.

**Comment by erlend-aasland** [comment]
Please, when adding the `pending` label, accompany it with a comment that explains the reason for applying the label. We cannot read your thoughts; it is better when you write them down for us.

File: .coveragerc
Comment (terryjreedy) [inline .coveragerc]:
```suggestion
    .*# pragma: no cover
    .*# pragma: no branch
```
Since I don't know exactly how these regexes are used, I don't know if these additions makes any difference or not.  But they match what I have, which may have been copied or left as is.  They do make clear that a line with anything followed by these comments should match and be excluded.

File: .coveragerc
Comment (terryjreedy) [inline .coveragerc:4]:
```suggestion
[run]
branch = True

[report]
```
I run IDLE tests with 'branch = True' and consider this the proper choice.  I don't know if others dispute this or not.  OP Ammar Askar, in https://github.com/python/cpython/pull/8150#issuecomment-403193951, removed this option because this "caused some reliably reproducible test errors".  He did not list them or get back to "make bpo bugs for them later after investigating".  This puzzles me since this option should only affect the coverage report, not the test execution itself.  

Ammar's later comment https://github.com/python/cpython/pull/8150#issuecomment-536350378 seems relevant but I don't understand what he wrote.  He appears to say that codecov would pay attention to this file placed where proposed, but I know nothing about codecov or whether the linked outputs were run with this file and if so with or without the branch option.

In any case, I believe I added DO NOT MERGE because this branch option issue should be resolved first.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Since CI is green without `branch = True`, let us see what happens now with it added.

EDIT: If codecov is not run even unofficially (not sure), this will not do any good.  I am running tests sequentially on a fresh local build and will try to manually run with my .coveragerc.

**Review by terryjreedy — APPROVED** [review]

**Comment by terryjreedy** [comment]
@Yhg1s Should I backport this to 3.12?  It is a new feature, but does not affect any code that does not run coverage or look at repository directory.

**Comment by erlend-aasland** [comment]
@terryjreedy, sorry for intruding, but I took the liberty to rename the NEWS file; it used the PR, not the issue, number in its file name.

File: .coveragerc
Comment (sobolevn) [inline .coveragerc:8]: It is a good idea to add:
- `raise NotImplementedError`
- `r'^ +\.\.\.$',`
- `r'^\s*if (False|TYPE_CHECKING):',`

See https://github.com/asottile/covdefaults for more useful ideas.

Comment (terryjreedy) [reply .coveragerc:8]: This file is currently the coverage default + IDLE additions, designed to be unlikely to affect other files (but not guaranteed, I admit, which is why I labelled them) + checking branches both ways.  Anything added should be something that should always be excluded from a whole-suite coverage test.

* NotImplementedError: It seems to me that test of abstract methods might use `assertRaises(NotImplentedError, ...)`.
* Re appears to match ` ...` line.  How might this be used in code and why would this never by executed by any test?
* False == 0 so I will add that if @erlend-aasland agrees.  Perhaps should be in coverage default though `if 0:` is traditional and easier to add for temporary disablement, which is the only sensible use of this construct.
* Please explain `TYPE_CHECKING` and its use.  (Being all caps it should be a constant, always true or always false, and thus not suitable as a conditional.)

**Comment by terryjreedy** [comment]
@erlend-aasland Intrude? Thank you for the fix.  I just opened the issue to enable that.

**Comment by terryjreedy** [comment]
Bogus failure in required test: test_threading changed environment
https://github.com/python/cpython/actions/runs/5536029870/jobs/10103154036?pr=8150

File: .coveragerc
Comment (erlend-aasland) [reply .coveragerc:8]: I think the current PR is a good starting point; IMO we should land it pretty much as it is.

We can discuss and cherry-pick regexes from Anthony's config in a follow up PR. For example, I'm not sure about the NotImplemented case for the reasons Terry provided.

**Pull request merged**
